components/target_estimator.py

```diff
@@ -0,0 +1,142 @@
+from components.chassis import Chassis
+from components.range_finder import RangeFinder
+from components.turret import Turret
+from components.vision import Vision, VisionData
+
+from utilities.functions import constrain_angle
+
+import math
+import time
+
+from typing import Optional
+
+from magicbot import feedback
+
+
+def _test_for_downrange(heading: float, azimuth: float) -> bool:
+    # Calculate the turret azimuth in field coordinates
+    turret_in_field = constrain_angle(heading + azimuth)
+    return abs(turret_in_field) < math.pi / 2.0
+
+
+class TargetEstimator:
+
+    ALPHA = 0.1  # Exponential smoothing fraction
+    CAMERA_TO_LIDAR = 0.15
+
+    chassis: Chassis
+    range_finder: RangeFinder
+    turret: Turret
+    vision: Vision
+
+    def __init__(self) -> None:
+        self._init()
+
+    def _init(self):
+        self.angle_to_target: Optional[float] = None
+        self.vision_range: Optional[float] = None
+        self.previous_azimuth: Optional[float] = None
+        self.previous_heading: Optional[float] = None
+        self.previous_vision_data: Optional[VisionData] = None
+        self._pointing_downrange: Optional[bool] = None
+
+    def on_enable(self) -> None:
+        self.reset()
+
+    def reset(self) -> None:
+        self._init()
+        heading = self.chassis.get_heading()
+        azimuth = self.turret.get_azimuth()
+        self._pointing_downrange = _test_for_downrange(heading, azimuth)
+        self.previous_heading = heading
+        self.previous_azimuth = azimuth
+
+    @feedback
+    def is_ready(self) -> bool:
+        return (
+            self.is_pointing_downrange is not None
```

Comment:
Isn't `self.is_pointing_downrange` a method?

Reply:
Indeed it is. The member is _pointing_dowrange. Fixed.